src/test/scala/mesosphere/marathon/core/task/tracker/impl/InstancesLoaderImplTest.scala

```diff
@@ -61,5 +72,39 @@ class InstancesLoaderImplTest extends AkkaUnitTest {
       val expectedData = InstanceTracker.InstancesBySpec.forInstances(app1Instance1, app1Instance2, app2Instance1)
       loaded.futureValue should equal(expectedData)
     }
+
+    "ignore instances without a run spec" in {
+      val f = new Fixture
+
+      Given("instances for multiple runSpecs")
+      val app1Id = PathId("/app1")
+      val app1Instance1 = TestInstanceBuilder.newBuilder(app1Id).getInstance()
+      val app1Instance2 = TestInstanceBuilder.newBuilder(app1Id).getInstance()
+      val app1 = app1Instance1.runSpec
+      f.groupRepository.runSpecVersion(eq(app1Id), eq(app1.version.toOffsetDateTime))(any) returns Future.successful(Some(app1))
+
+      val app2Id = PathId("/app2")
+      val app2Instance1 = TestInstanceBuilder.newBuilder(app2Id).getInstance()
+      f.groupRepository.runSpecVersion(eq(app2Id), eq(app2Instance1.runSpecVersion.toOffsetDateTime))(any) returns Future.successful(None)
+
+      val instances = Seq(
+        StateInstance.fromCoreInstance(app1Instance1),
+        StateInstance.fromCoreInstance(app1Instance2),
+        StateInstance.fromCoreInstance(app2Instance1)
+      )
+
+      f.instanceRepository.ids() returns Source(instances.map(_.instanceId)(collection.breakOut))
+      for (instance <- instances) {
+        f.instanceRepository.get(instance.instanceId) returns Future.successful(Some(instance))
+      }
+
+      When("load is called")
+      val loaded = f.loader.load()
+
+      Then("the resulting data ")
+      // we do not need to verify the mocked calls because the only way to get the data is to perform the calls
+      val expectedData = InstanceTracker.InstancesBySpec.forInstances(app1Instance1, app1Instance2)
```

Comment:
So without the correct _runSpec_ and _version_ the instance will *not be part of the state*... until a _StatusUpdate_ for it is received and it will be expunged (because its unknown)? But we won't actively reconcile for it since it is not part of the `instancesBySpec` call? So In a case, when Mesos doesn't know the task, we'll end up with some permanent garbage in ZK?

Reply:
> So In a case, when Mesos doesn't know the task, we'll end up with some permanent garbage in ZK?

You mean Mesos forgets about the instance for some reason and Marathon forgets about the run spec for some reason and then we don't load it but keep it in ZooKeeper? Well, the bug is that we should never have instances without a run spec to begin with. That is a bug. Even in the current version we always should have the run spec. We do handle that situation by just not scaling https://github.com/mesosphere/marathon/blob/master/src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala#L492. That means we never start a new instance for a failed if we do not know the run spec.

What would be the best behaviour here? Current Marathon will know the instance but not the run spec and thus won't be able to handle it. What do you think about `case object Unknown extends RunSpec`?

Reply:
No, I meant, if there is an instance but for some reason, `RunSpec` was deleted. Do you think we should we log the problem and expunge the Instance from the state?

Reply:
Ah, that would be simpler but might be unexpected behaviour since Marathon will kill that instance which it would do with this change as well.

If we do not load that instance we should delete it from the state as well. If we keep it in state we should load it with an `UnknownRunSpec`.

Reply:
So we either:
a) Expunge the instance from the state right away and change how we deal with orphaned tasks - we shouldn't kill them just because we lost their state or
b) Keep the instance with `UnknownRunSpec` in the state and deal with implications or e.g. rescheduling Instances in the scheduling/orchestrator layer